src/managers/QueueManager.sol

```diff
@@ -0,0 +1,113 @@
+
+    /// @inheritdoc IUpdateContract
+    function update(PoolId poolId, ShareClassId scId, bytes calldata payload) external {
+        require(msg.sender == contractUpdater, NotContractUpdater());
+
+        uint8 kind = uint8(UpdateContractMessageLib.updateContractType(payload));
+        if (kind == uint8(UpdateContractType.UpdateQueue)) {
+            UpdateContractMessageLib.UpdateContractUpdateQueue memory m =
+                UpdateContractMessageLib.deserializeUpdateContractUpdateQueue(payload);
+            ShareClassQueueState storage sc = scQueueState[poolId][scId];
+            sc.minDelay = m.minDelay;
+            sc.extraGasLimit = m.extraGasLimit;
+            emit UpdateQueueConfig(poolId, scId, m.minDelay, m.extraGasLimit);
+        } else {
+            revert UnknownUpdateContractType();
+        }
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Sync
+    //----------------------------------------------------------------------------------------------
+
+    /// @inheritdoc IQueueManager
+    function sync(PoolId poolId, ShareClassId scId, AssetId[] calldata assetIds) external {
+        ShareClassQueueState storage sc = scQueueState[poolId][scId];
+        require(
+            sc.lastSync == 0 || sc.minDelay == 0 || block.timestamp >= sc.lastSync + sc.minDelay, MinDelayNotElapsed()
+        );
+
+        (uint128 delta,, uint32 queuedAssetCounter,) = balanceSheet.queuedShares(poolId, scId);
+        require(delta > 0 || queuedAssetCounter > 0, NoUpdates());
+
+        require(assetIds.length <= 256, TooManyAssets()); // Bitmap limit
+
+        // Deduplicate and validate using bitmap for valid indices
+        uint256 validBitmap = 0; // Each bit represents if that index is valid
+        uint256 validCount = 0;
+
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            bytes32 key = bytes32(uint256(AssetId.unwrap(assetIds[i])));
+            if (TransientStorageLib.tloadBool(key)) {
+                continue; // Skip duplicate
+            }
+            TransientStorageLib.tstore(key, true);
+
+            // Check if valid
+            (uint128 deposits, uint128 withdrawals) = balanceSheet.queuedAssets(poolId, scId, assetIds[i]);
+            if (deposits > 0 || withdrawals > 0) {
+                validBitmap = validBitmap.withBit(i, true);
+                validCount++;
+            }
+        }
+
+        // Build exactly-sized array
+        bool submitShares = delta > 0 && validCount >= queuedAssetCounter;
+        bytes[] memory cs = new bytes[](validCount + (submitShares ? 1 : 0));
+
+        uint256 csIndex = 0;
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            if (validBitmap.getBit(i)) {
+                // Check if index i is valid
+                cs[csIndex++] = abi.encodeWithSelector(
+                    balanceSheet.submitQueuedAssets.selector, poolId, scId, assetIds[i], sc.extraGasLimit
```

Comment:
Theoretically, extraGasLimit should be a parameter of `sync()`. 

The explanation: although the snapshot hook can be the same over time, how the snapshot hook behaves internally based on the message could differ a lot. Putting this value in the `sync()` method, allows the caller to play with that value in cases where some extra computation requires more gas

Reply:
I think the pool manager should just set the worst case `extraGasLimit` then

Problem with making it a parameter is that this method is permissionless, so then callers can just set 0, and it will still fail with out of gas on the target chain.